### DISCUSSION TITLE
working with custom csv data

### DISCUSSION THREAD

**Discussion by chillum-codeX** [pyg-team/pytorch_geometric#3844] [Q&A]
working with custom csv data
i have 12 features and have to predict seriousness

**Comment by rusty1s** [comment]
What's the question? :)

**Reply by chillum-codeX** [reply]
[dummy data.xlsx](https://github.com/pyg-team/pytorch_geometric/files/7886839/dummy.data.xlsx)
this was the dataset
I have to predict the seriousness
we have multiple columns 
in symptoms we have different and many adverse events 

how can i create nodes for each symptom
what model should i use for predicting seriousness in hetrodata
how can i create individual nodes for each symptom and how should i visualize this hetrodata

DATA SOURCE >> https://vaers.hhs.gov/data/datasets.html

This is our sample dataset. We want to predict the column LTHREAT what kind of model is useful for this prediction, we also have n number of symptoms in different columns and we need to create individual nodes for each unique symptom how can we do that and how to visualize this heterogeneous graph

**Reply by rusty1s** [reply]
In general, each row is interpreted as a node/example in your dataset, and each column represents a feature that needs to be converted to a numerical representation and added to `data.x`. However, there does not seem to exist any notion of an edge in your dummy table. How do you want to connect different nodes/rows with each other?

**Reply by chillum-codeX** [reply]
My edges dont have any weights 
how should i create square matrix

Our nodes dont have any features and our edges dont have any weights. As we dont have any features for the node we don't have node.x tensors  so how to deal with it.

In movieLens example they have created a square matrix(one hot encodings) for the node without feature. Do we need to follow the same or is there any other way to define the node

**Reply by rusty1s** [reply]
I'm sorry but I'm still not sure I understand. Can you give an overview of the different tables and schemas you have in your dataset? If you do not have any node features, you need to create them since they are necessary as input to a GNN. One option is indeed to utilize an identity matrix, another option is to utilize `torch.nn.Embedding` to embed each node prior to GNN execution.

**Reply by chillum-codeX** [reply]
![2022-01-20_12h19_57](https://user-images.githubusercontent.com/40481821/150288143-fd7be736-0b30-4b48-8e91-8d59a3267e5a.png)

Whenever a patient takes a vaccine and experiences any adverse event then he reports it and a case ID(VAERS_ID) is created.



Nodes like symptoms experienced, other medicines he took, serioudness of the event(LTHREAT : life threatening situation) etc are considered with repective relationship as shown in the image.



Our aim is to predict whether a case will lead to life threatening(LTHREAT) situation or not on the basis of symptoms experienced, vaccine taken and other medicines the patient took.



Here nodes like symptom, other medicines, vaccine and Lthreat doesn't have any features and edges are unweighed. So how will we define the nodes without features and unweighed edges.

**Reply by rusty1s** [reply]
Thank you. While I think one can interpret this as a graph, isn't it more naturally to encode these features (symptons, age, gender, ...) into a single feature vector and utilize traditional machine learning with IID samples?

Alternatively, you can interpret each symptom, etc as a node in your graph, and learn an embedding for that node in an end-to-end fahsion jointly with the prediction task, e.g., via `torch.nn.Embedding`. Happy to chat more about this.

**Reply by chillum-codeX** [reply]
Instead of using nn.embeddings to define a node without features is there any other way to define it

**Reply by rusty1s** [reply]
Yes, there exists. One alternative way is to define node features by some graph statistics, e.g., node degree. We provide support for this via [`transforms.OneHotDegree`](https://pytorch-geometric.readthedocs.io/en/latest/modules/transforms.html#torch_geometric.transforms.OneHotDegree) or [`transforms.LocalDegreeProfile`](https://pytorch-geometric.readthedocs.io/en/latest/modules/transforms.html#torch_geometric.transforms.LocalDegreeProfile).

**Reply by chillum-codeX** [reply]
Thanks @rusty1s 
i will try and let you know
